lib/chef-dk/policyfile/git_lock_fetcher.rb

```diff
@@ -0,0 +1,262 @@
+    #
+    # @since 3.1.0
+    #
+    class GitLockFetcher
+      attr_accessor :name
+      attr_accessor :source_options
+      attr_accessor :storage_config
+
+      attr_reader :uri
+      attr_reader :revision
+      attr_reader :path
+      attr_reader :branch
+      attr_reader :tag
+      attr_reader :ref
+
+      # Initialize a GitLockFetcher
+      #
+      # @param name [String] The name of the policyfile
+      # @param source_options [Hash] A hash with a :path key pointing at the location
+      #                              of the lock
+      def initialize(name, source_options, storage_config)
+        @name           = name
+        @storage_config = storage_config
+        @source_options = symbolize_keys(source_options)
+        @revision = @source_options[:revision]
+        @path     = @source_options[:path] || @source_options[:rel]
+        @uri      = @source_options[:git]
+        @branch   = @source_options[:branch]
+        @tag      = @source_options[:tag]
+        @ref      = @source_options[:ref]
+
+        # The revision to parse
+        @rev_parse = @source_options[:ref] || @source_options[:branch] || @source_options[:tag] || "master"
+      end
+
+      # @return [True] if there were no errors with the provided source_options
+      # @return [False] if there were errors with the provided source_options
+      def valid?
+        errors.empty?
+      end
+
+      # Check the options provided when craeting this class for errors
+      #
+      # @return [Array<String>] A list of errors found
+      def errors
+        error_messages = []
+        [:git].each do |key|
+          error_messages << "include_policy for #{name} is missing key #{key}" unless source_options[key]
+        end
+
+        error_messages
+      end
+
+      # @return [Hash] The source_options that describe how to fetch this exact lock again
+      def source_options_for_lock
+        source_options.merge({
+                               revision:  revision,
+                             })
+      end
+
+      # Applies source options from a lock file. This is used to make sure that the same
+      # policyfile lock is loaded that was locked
+      #
+      # @param options_from_lock [Hash] The source options loaded from a policyfile lock
+      def apply_locked_source_options(options_from_lock)
+        options = options_from_lock.inject({}) do |acc, (key, value)|
+          acc[key.to_sym] = value
+          acc
+        end
+        source_options.merge!(options)
+        raise ChefDK::InvalidLockfile, "Invalid source_options provided from lock data: #{options_from_lock_file.inspect}" if !valid?
+      end
+
+      # @return [Hash] of the policyfile lock data
+      def lock_data
+        @lock_data ||= fetch_lock_data.tap do |data|
+          data["cookbook_locks"].each do |cookbook_name, cookbook_lock|
+            cookbook_lock["source_options"].tap do |opt|
+              opt["git"]      = uri unless opt.has_key?("git")
```

Comment:
I'm not sure about this. This logic only seems valid if the cookbook is a local cookbook, but there's no check for that

Reply:
here's just a piece of the lock data. It looksl ike it gets conflicting information, however happens to work (probably based on the ordering in which things are selected):
```
 "cookbook_locks"=>
  {"chef_handler"=>
    {"version"=>"3.0.2",
     "identifier"=>"60cbc4e26770cf157c9f5ec71c0abfdeb79784e5",
     "dotted_decimal_identifier"=>"27245644236681423.6047998442675210.210963283805413",
     "cache_key"=>"chef_handler-3.0.2-supermarket.chef.io",
     "origin"=>"https://supermarket.chef.io:443/api/v1/cookbooks/chef_handler/versions/3.0.2/download",
     "source_options"=>
      {"artifactserver"=>
        "https://supermarket.chef.io:443/api/v1/cookbooks/chef_handler/versions/3.0.2/download",
       "version"=>"3.0.2",
       "git"=>"https://github.com/jaym/policyfile-book",
       "revision"=>"971faffb36f0b42e233e576a6ece3ea93133609d",
       "branch"=>"master",
       "tag"=>nil}},
```